rfcs/splittable-prng.md

```diff
@@ -0,0 +1,103 @@
+
+### Concern: good randomness
+
+We want PRNGs that pass statistical tests for good randomness. All algorithms discussed here perform well under standard PRNG-quality test (diehard, etc.).
+
+
+### Possible concern: lesser portability of C stubs
+
+Random currently has a pure-OCaml implementation (except for the system-specific auto-seeding logic). Moving to C stubs may be an issue, at least for Mirage users, and require extra work from alternative implementations (js_of_ocaml, BuckleScript, etc.).
+
+SplitMix is a very simple algorithm, it is easy to provide a pure-OCaml version that should perform well. Porting Chacha, which is more elaborate, is more work, but it's also harder to predict performance for the pure-OCaml version.
+
+
+### Possible concern: Javascript performance
+
+js_of_ocaml and Bucklescript/ReScript are important for our users, and it would be nice to ensure that the performance is not disappointing on these alternative implementations.
+
+js_of_ocaml implements int64 with an emulation, suggesting that SplitMix could take a performance hit. (There is no native int64 type under Javascript anyway, "mainstream" approaches such as [long.js](https://github.com/dcodeIO/long.js) emulate them with two numbers.) It may be that JS engines optimize the emulation layer efficiently, so we should evaluate each choice.
+
+(Using C stubs in fact gives more flexibility for alternative implementations to provide their own native implementation of these functions.)
+
+
+## Summary
+
+What are the requirements for the Random module?
+
+The requirements discussed in the current version of the RFC are:
+- implementation being pure OCaml
+- performance
+- "security" of the PRNG
+- reseeding recommendations
+
+
+Here is a summary of what we understand to be the consensus so far (I'll edit this part of the RFC as discussion progresses):
+
+- pure-OCaml: yes. 
+
+  A pure-OCaml implementation makes people's life simpler would be strongly preferable; we should port SplitMix and Chacha and re-run benchmarks.
+
+- performance: not important, but check that js_of_ocaml does okay.
+
+  A small performance hits are not very important for the standard Random PRNG, so a reasonable decrease (for everyone or just js_of_ocaml) would be perfectly acceptable.
+  
+  We should still benchmark under js_of_ocaml to have an idea of the performance (once we have pure-OCaml versions). It's bad if some parts of OCaml programs suddenly become much slower there.
+
+- Having a crypto-secure PRNG is not part of our requirement -- Random is not secure in that sense.
+
+- Having to reseed SplitMix every 2^30 draws may be a problem in practice -- loops running 2^30 iterations are perfectly reasonable thse days. All other things being equal, ChaCha should be preferred.
```

Comment:
Again it's 2^32.  "Having to reseed" may be too strong: it is recommended to reseed after 2^32 draws.  Basically, with a N-bit counter as internal state, some statistical deviations are to be expected after sqrt(2^N) numbers have been generated, because of the birthday paradox and all that.  But again I don't have a reference handy.